server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java

```diff
@@ -191,16 +210,18 @@ public SegmentReplicationState getSegmentReplicationState(ShardId shardId) {
      */
     public synchronized void onNewCheckpoint(final ReplicationCheckpoint receivedCheckpoint, final IndexShard replicaShard) {
         logger.trace(() -> new ParameterizedMessage("Replica received new replication checkpoint from primary [{}]", receivedCheckpoint));
+        // if the shard is in any state
+        if (replicaShard.state().equals(IndexShardState.CLOSED)) {
+            // ignore if shard is closed
+            logger.trace(() -> "Ignoring checkpoint, Shard is closed");
+            return;
+        }
+        updateLatestReceivedCheckpoint(receivedCheckpoint, replicaShard);
```

Comment:
nit: Can you add an integration test verifying the fix ? One way would be to block recovery on replica while primary performs indexing.

Reply:
Added test in latest commit.